### PR TITLE
Rebase of updating UART to match TRD

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request is a rebase of #3046.

Is this something we are still pursuing? The original PR is very minimal. This version remove the additional boards and a lot of commented out code.

### Testing Strategy

todo

### TODO or Help Wanted

Much

Also if anyone cares about the commits please fix them. I just squashed most of everything from the original PR.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

**Comment by hudson-ayers** [comment]
Something definitely got mixed up in that PR at some point across multiple people pushing to it...the sam4l uart implementation in this PR doesn't do anything, whereas we had sam4l fully ported to the new interface and tested and working. I will try to figure out what happened here

**Comment by TristenSeth** [comment]
I have a branch on Jay's fork that I just reviewed. It has all the proper code for the NRF52 UART, and the required USB capsule stuff. It also has the Segger RTT stuff which I think is already in this commit history (thanks Colin!), so feel free to cherry pick that I guess.

Here is the [branch](https://github.com/jaehyeonjay/tock/tree/trd-uart-nrf52)

Let me know if there is anything else I can do to help.

Please do try and find a way to merge this while preserving the commit history, so the contribution can appear on all the contributors profiles (Somehow I was erased from the commit history on the original PR, despite having done a majority of the NRF52 updates.) No biggie, just something I though I would mention.

**Comment by TristenSeth** [comment]
> I have a branch on Jay's fork that I just reviewed. It has all the proper code for the NRF52 UART, and the required USB capsule stuff. It also has the Segger RTT stuff which I think is already in this commit history (thanks Colin!), so feel free to cherry pick that I guess.
> 
> Here is the [branch](https://github.com/jaehyeonjay/tock/tree/trd-uart-nrf52)
> 
> Let me know if there is anything else I can do to help.
> 
> Please do try and find a way to merge this while preserving the commit history, so the contribution can appear on all the contributors profiles (Somehow I was erased from the commit history on the original PR, despite having done a majority of the NRF52 updates.) No biggie, just something I though I would mention.

I went back to this today after some sleep since my CI tests failed. It seems what the problem is was that we had too many similarly named branches. Specifically, by the end we had two branches, one named 'trd-uart' and one named 'trd-uart-final'. Somehow, the 'trd-uart-final' branch never got the NRF52 code merged, nor the USB capsule changes I made along with that. The 'trd-uart' branch actually had both the SAM4L and NRF52 updates, as well as both the Segger RTT and USB capsule updates. I went ahead and got everything set correctly (I think) for both the SAM4L and NRF52 UART implementations, as well as the USB and SeggerRTT capsules. Since this pull request already has the Sifive commit for it's uart, I did not merge that with the 'trd-uart' branch. 

So, just to clarify. I believe that the correct branch with all of the new HIL implementation for the SAM4L and NRF52 and the capsule updates, are on Jaeh's tock fork, on the 'trd-uart' branch (**not** the 'try-uart-final'!)

I hope this straightens it all out! Let me know if there is anything I can help with since I know there have been a ton of changes committed since May (when we pushed these).

**Comment by lschuermann** [comment]
@bradjc To me this, at least superficially, seems like something we still want to have (or, at the very least eventual consistency between our documentation and implementation). It's been quite a while, but while porting over the 1-2 UART implementations I tackled back then, the new HIL made sense to me.

Should we bring this up for discussion on the next core call?

**Comment by bradjc** [comment]
> Should we bring this up for discussion on the next core call?

What are we discussing? I doubt anyone is opposed, just need someone to actually figure this PR out and get it across the finish line.

**Comment by lschuermann** [comment]
@bradjc Fair enough. If I recall correctly, our strategy here was to have board maintainers migrate their implementations to the new HIL. I'm happy to do my part and migrate a couple, and then we can see how to drive this further. Clearly, the old approach of waiting until it "magically happens" isn't working, so we should have a forcing function (e.g., blocking the next release on this).

**Comment by lschuermann** [comment]
I tried rebasing these changes, however the codebase has moved so much (even in UART implementations like SiFive) that it will likely be easier to port take the HIL changes, and re-port the chips that were already migrated.

> Please do try and find a way to merge this while preserving the commit history, so the contribution can appear on all the contributors profiles 

@colin-r-schultz I will likely need to re-create the original commit that performed the HIL changes, and split out all of the other unrelated changes. I can add a `Co-authored-by` trailer to preserve authorship provenance.

File: kernel/src/hil/uart.rs [hunk 1]
```diff
@@ -117,7 +147,7 @@ pub trait Transmit<'a> {
     /// Transmit a single word of data asynchronously. The word length is
     /// determined by the UART configuration: it can be 6, 7, 8, or 9 bits long.
     /// If the `Result<(), ErrorCode>` is Ok(()), on completion,
-    /// `transmitted_word` will be called on the `TransmitClient`.
+    /// `transmitted_character` will be called on the `TransmitClient`.
     /// Other valid `Result<(), ErrorCode>` values are:
     ///  - OFF: The underlying hardware is not available, perhaps because
```
Comment (wprzytula) [inline kernel/src/hil/uart.rs:152]: Why this change of naming? I associate `character` strictly with one ASCII character, so 8 bits. OTOH, `word` has semantics of being strictly dependent on the hardware, which is exactly the case here.

Comment (wprzytula) [reply kernel/src/hil/uart.rs:152]: WDYT about making this strongly typed in a way that `Word` becomes an enum:
```rust
enum Word {
  Len6(u16),
  Len7(u16),
  Len8(u16),
  Len9(u16),
}
```
so that it becomes clearer for the HIL users that only a couple of lower bits of `word` argument is actually passed to the hardware, even though the type is much bigger (`u32` in the current code, `u16` as I propose, for the whole tagged enum to fit in a single `u32`).

Comment (lschuermann) [reply kernel/src/hil/uart.rs:152]: I think those are both good suggestions, but I think those should go into the UART TRD itself which specifies this interface. Feel free to create a PR for them!

**Comment by lschuermann** [comment]
Closed in favor of #4181 as discussed.
